scripts/validate_package_entries/validate_package_entries.py

```diff
@@ -0,0 +1,159 @@
+#!/usr/bin/env python3
+"""Validate that package entries in pyproject.toml are up to date.
+
+This script discovers all Python packages in the components/ and pipelines/
+directories and ensures they are properly listed in pyproject.toml under
+tool.setuptools.packages.
+
+Usage:
+    uv run python scripts/validate_package_entries/validate_package_entries.py
+"""
+
+import argparse
+import sys
+import tomllib
+from pathlib import Path
+from typing import Set
+
+
+def get_repo_root() -> Path:
+    """Get the repository root directory."""
+    return Path(__file__).parent.parent.parent.resolve()
+
+
+def discover_packages(repo_root: Path) -> Set[str]:
+    """Discover all Python packages in components/ and pipelines/ directories.
+
+    Returns a set of package names in the format kfp_components.* based on
+    the package-dir mapping in pyproject.toml.
+    """
+    packages: Set[str] = set()
+
+    # Always include the root package
+    if (repo_root / "__init__.py").exists():
+        packages.add("kfp_components")
+
+    def _discover_recursive(directory: Path, base_package: str) -> None:
+        """Recursively discover packages in a directory."""
+        if not directory.exists():
+            return
+
+        for item in directory.iterdir():
+            if item.is_dir() and (item / "__init__.py").exists():
+                # Convert path to package name
+                rel_path = item.relative_to(repo_root)
+                package_name = "kfp_components." + ".".join(rel_path.parts)
+                packages.add(package_name)
+
+                # Recursively discover nested packages
+                _discover_recursive(item, package_name)
+
+    # Discover packages in components/
+    components_dir = repo_root / "components"
+    if components_dir.exists() and (components_dir / "__init__.py").exists():
+        packages.add("kfp_components.components")
+        _discover_recursive(components_dir, "kfp_components.components")
+
+    # Discover packages in pipelines/
+    pipelines_dir = repo_root / "pipelines"
+    if pipelines_dir.exists() and (pipelines_dir / "__init__.py").exists():
+        packages.add("kfp_components.pipelines")
+        _discover_recursive(pipelines_dir, "kfp_components.pipelines")
+
+    return packages
+
+
+def read_pyproject_packages(repo_root: Path) -> Set[str]:
+    """Read the packages list from pyproject.toml."""
+    pyproject_path = repo_root / "pyproject.toml"
+
+    try:
+        with open(pyproject_path, "rb") as f:
+            pyproject = tomllib.load(f)
+    except FileNotFoundError:
+        raise RuntimeError(f"pyproject.toml not found at {pyproject_path}")
+    except tomllib.TOMLDecodeError as e:
+        raise RuntimeError(f"Failed to parse pyproject.toml: {e}")
```

Comment:
The `tomllib.TOMLDecodeError` is caught and re-raised as a `RuntimeError`, which loses the original traceback context.

```suggestion
        raise RuntimeError(f"Failed to parse pyproject.toml: {e}") from e
```